/**
 * ComponentInfo.java 10:36:09 PM Apr 21, 2008
 * 
 * <PRE>
 * Copyright (c) 2008, Jan Amoyo
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 *     - Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     - Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 'AS IS';
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
 * POSSIBILITY OF SUCH DAMAGE.
 * </PRE>
 */

package org.fixsuite.message.info;

import org.checkerframework.dataflow.qual.Impure;
import org.checkerframework.dataflow.qual.Pure;
import org.checkerframework.dataflow.qual.SideEffectFree;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.TreeMap;

/**
 * Represents a Component as defined by the FIX specification.
 * 
 * @author jramoyo
 */
public class ComponentInfo implements CompositeFixInfo {

    private String name;

    private String componentType;

    private String category;

    private int id;

    private String abbreviation;

    private boolean isNotRequiredXml;

    private boolean isRequired;

    // Default collection
    private TreeMap<Double, FixInfo> itemsByPosition;

    private HashMap<FixInfo, Double> itemsByContent;

    // Default collection
    private TreeMap<Integer, FieldInfo> fieldsByTagNumber;

    private TreeMap<String, FieldInfo> fieldsByName;

    // Default collection
    private TreeMap<String, ComponentInfo> componentsByName;

    private TreeMap<Integer, ComponentInfo> componentsById;

    /**
     * @see org.fixsuite.message.info.CompositeFixInfo#getFields()
     */
    @SideEffectFree
    public List<FieldInfo> getFields() {
        if (fieldsByTagNumber != null) {
            return new ArrayList<FieldInfo>(fieldsByTagNumber.values());
        }
        return null;
    }

    /**
     * @see org.fixsuite.message.info.CompositeFixInfo#getField(int)
     */
    @Pure
    public FieldInfo getField(int tagNumber) {
        if (fieldsByTagNumber != null) {
            return fieldsByTagNumber.get(tagNumber);
        }
        return null;
    }

    /**
     * @see org.fixsuite.message.info.CompositeFixInfo#getField(java.lang.String)
     */
    @Pure
    public FieldInfo getField(String name) {
        if (fieldsByName != null) {
            return fieldsByName.get(name);
        }
        return null;
    }

    /**
     * @see org.fixsuite.message.info.CompositeFixInfo#addField(org.fixsuite.message.info.FieldInfo,
     *      double)
     */
    @Impure
    public void addField(FieldInfo field, double position) {
        if (fieldsByTagNumber == null) {
            fieldsByTagNumber = new TreeMap<Integer, FieldInfo>();
            fieldsByName = new TreeMap<String, FieldInfo>();
        }
        fieldsByTagNumber.put(field.getTagNumber(), field);
        fieldsByName.put(field.getName(), field);
        addItem(field, position);
    }

    /**
     * @see org.fixsuite.message.info.CompositeFixInfo#getComponents()
     */
    @SideEffectFree
    public List<ComponentInfo> getComponents() {
        if (componentsByName != null) {
            return new ArrayList<ComponentInfo>(componentsByName.values());
        }
        return null;
    }

    /**
     * @see org.fixsuite.message.info.CompositeFixInfo#getComponent(int)
     */
    @Pure
    public ComponentInfo getComponent(int id) {
        if (componentsById != null) {
            return componentsById.get(id);
        }
        return null;
    }

    /**
     * @see org.fixsuite.message.info.CompositeFixInfo#getComponent(java.lang.String)
     */
    @Pure
    public ComponentInfo getComponent(String name) {
        if (componentsByName != null) {
            return componentsByName.get(name);
        }
        return null;
    }

    /**
     * @see org.fixsuite.message.info.CompositeFixInfo#addComponent(org.fixsuite.message.info.ComponentInfo,
     *      double)
     */
    @Impure
    public void addComponent(ComponentInfo component, double position) {
        if (componentsByName == null) {
            componentsByName = new TreeMap<String, ComponentInfo>();
            componentsById = new TreeMap<Integer, ComponentInfo>();
        }
        componentsByName.put(component.getName(), component);
        componentsById.put(component.getId(), component);
        addItem(component, position);
    }

    /**
     * @see org.fixsuite.message.info.CompositeFixInfo#getItems()
     */
    @SideEffectFree
    public List<FixInfo> getItems() {
        if (itemsByPosition != null) {
            return new ArrayList<FixInfo>(itemsByPosition.values());
        }
        return null;
    }

    /**
     * @see org.fixsuite.message.info.CompositeFixInfo#replaceAsGroup(org.fixsuite.message.info.FieldInfo,
     *      org.fixsuite.message.info.GroupInfo)
     */
    @Impure
    public void replaceAsGroup(FieldInfo field, GroupInfo group) {
        fieldsByTagNumber.put(field.getTagNumber(), group);
        fieldsByName.put(field.getName(), group);
        double position = itemsByContent.get(field);
        itemsByContent.remove(field);
        itemsByContent.put(group, position);
        itemsByPosition.put(position, group);
    }

    /**
     * Returns the name
     * 
     * @return the name
     */
    @Pure
    public String getName() {
        return name;
    }

    /**
     * Modifies the name
     * 
     * @param name - the name to set
     */
    @Impure
    public void setName(String name) {
        this.name = name;
    }

    /**
     * Returns the componentType
     * 
     * @return the componentType
     */
    @Pure
    public String getComponentType() {
        return componentType;
    }

    /**
     * Modifies the componentType
     * 
     * @param componentType - the componentType to set
     */
    @Impure
    public void setComponentType(String componentType) {
        this.componentType = componentType;
    }

    /**
     * Returns the category
     * 
     * @return the category
     */
    @Pure
    public String getCategory() {
        return category;
    }

    /**
     * Modifies the category
     * 
     * @param category - the category to set
     */
    @Impure
    public void setCategory(String category) {
        this.category = category;
    }

    /**
     * Returns the id
     * 
     * @return the id
     */
    @Pure
    public int getId() {
        return id;
    }

    /**
     * Modifies the id
     * 
     * @param id - the id to set
     */
    @Impure
    public void setId(int id) {
        this.id = id;
    }

    /**
     * Returns the abbreviation
     * 
     * @return the abbreviation
     */
    @Pure
    public String getAbbreviation() {
        return abbreviation;
    }

    /**
     * Modifies the abbreviation
     * 
     * @param abbreviation - the abbreviation to set
     */
    @Impure
    public void setAbbreviation(String abbreviation) {
        this.abbreviation = abbreviation;
    }

    /**
     * Returns the isNotRequiredXml
     * 
     * @return the isNotRequiredXml
     */
    @Pure
    public boolean isNotRequiredXml() {
        return isNotRequiredXml;
    }

    /**
     * Modifies the isNotRequiredXml
     * 
     * @param isNotRequiredXml - the isNotRequiredXml to set
     */
    @Impure
    public void setNotRequiredXml(boolean isNotRequiredXml) {
        this.isNotRequiredXml = isNotRequiredXml;
    }

    /**
     * Returns the isRequired
     * 
     * @return the isRequired
     */
    @Pure
    public boolean isRequired() {
        return isRequired;
    }

    /**
     * Modifies the isRequired
     * 
     * @param isRequired - the isRequired to set
     */
    @Impure
    public void setRequired(boolean isRequired) {
        this.isRequired = isRequired;
    }

    @Impure
    private void addItem(FixInfo item, double position) {
        if (itemsByPosition == null) {
            itemsByPosition = new TreeMap<Double, FixInfo>();
            itemsByContent = new HashMap<FixInfo, Double>();
        }
        itemsByPosition.put(position, item);
        itemsByContent.put(item, position);
    }

}
